libhdf5/CMakeLists.txt

```diff
@@ -28,7 +28,7 @@ if(STATUS_PARALLEL)
     target_link_libraries(netcdfhdf5 PUBLIC MPI::MPI_C)
 endif(STATUS_PARALLEL)
 
-target_include_directories(netcdfhdf5 PRIVATE ${HDF5_INCLUDE_DIR})
+target_include_directories(netcdfhdf5 PUBLIC ${HDF5_INCLUDE_DIR})
```

Comment:
I would suggest to draw out all of the targets. You can use [`--graphviz`](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-graphviz) tio get that.

About the linkage:
- `INTERFACE`: I don't use, next one does
- `PUBLIC`: I use, next one also does
- `PRIVATE`: I use, next one does not

For internal targets, you can choose whichever fits, but for the exported one, be very careful what is exported because that will appear in the `HDF5Config.cmake`.

Reply:
Thanks for this. I've been having trouble with getting hdf5 to be found by all targets, now that I've removed the global `include_directories`. I'll give the graphviz option a shot

Reply:
And I intended for `PUBLIC` here. Anything that links against `netcdfhdf5` needs access to hdf5, at least internal to the netcdf project.

Reply:
Yeah, the organization is quite messy, which is why I did not go into doing these when I initially wrote the issue. Much appreciate the effort you are putting here.

What would help here is to get some feedback from @WardF or other main devs on what the folder-structure is, specifically among the folders:
- which are supposed to be built like libraries that are meant to be installed: Make a proper `add_library`
- which are third-party libraries: Build as `add_library(STATIC)`
- which are simply sources: Add to an appropriate target

The folder structure should than be rearranged as:
```console
$ tree ./
./
├── include
├── src
│   ├── lib1
│   └── main
└── third_party
    └── dep1
```

Reply:
To answer the question, the only libraries we build/install (as final products) are the `libnetcdf` library and the filter interface libraries, located in `liblib/` and `plugins/`, respectively.  We also build executables, located in `ncdump/`, `ncgen/`, `ncgen3/`. The other directories prefaced with `lib` are components of the final `libnetcdf.a/libnetcdf.so`, and are gated by options at configure time. 

There aren't any `third_party` libraries to be installed, so there's no need for that.  What is the proposed impact on directories that are neither libraries (first or third party) or executables? The various testing directories, for example, or the docs directory, examples, etc?  

The second, fairly large concern I have is the fairly significant overhaul this would require to `configure.ac` and numerous `Makefile.am` files, to avoid breaking the `autoconf`-based build system.  This would constitute a not-insignficant effort, on top of the various changes to the cmake configuration. 

Finally, for better or worse, we have a fairly old and hierarchical set of shell scripts used for testing; they largely handle determining pathing, scripting, and plugin locations programmatically, and rely on relative pathing.  Changing the underlying layout in such a significant way *might* lead to some additional headaches trying to sort that out.  Alternatively, it might be fine; I'm not certain without digging into it.  @DennisHeimbigner any thoughts?

@K20shores my calendar is up to date if you'd like to discuss this via a video chat, or we can discuss it at our next regular meeting. I was going to set something up for Friday to check in, anyways. I'm not against the idea of cleaning up and reorganizing in a way that makes sense and lends itself to a cleaner layout and organization.  I just want to 1) Understand the benefits so that I can share an educated opinion on the cost/benefit in terms of effort, and 2) Communicate the potential broader impact of this work, as it would kick off a set of significant changes that go beyond the CMake build system.

Reply:
> There aren't any third_party libraries to be installed, so there's no need for that.

Third party is more related to dependencies that are included as-is or with modifications.

> What is the proposed impact on directories that are neither libraries (first or third party) or executables? The various testing directories, for example, or the docs directory, examples, etc?

Those are in `test`, `data`, `docs`, `examples`, `tools` directories, etc.. Preferably anything that does not relate to the project's build would be in its own directory. Other language bindings and anything that can be built as standalone would be its own directory as well, with a separate `project`. [Here](https://github.com/spglib/spglib/blob/develop/fortran/CMakeLists.txt)'s an example of that design.

> The second, fairly large concern I have is the fairly significant overhaul this would require to configure.ac and numerous Makefile.am files, to avoid breaking the autoconf-based build system. This would constitute a not-insignficant effort, on top of the various changes to the cmake configuration.
>
> Finally, for better or worse, we have a fairly old and hierarchical set of shell scripts used for testing; they largely handle determining pathing, scripting, and plugin locations programmatically, and rely on relative pathing. Changing the underlying layout in such a significant way might lead to some additional headaches trying to sort that out. Alternatively, it might be fine; I'm not certain without digging into it. @DennisHeimbigner any thoughts?

This is indeed a major chicken-and-egg issue. Downstream cannot change to cmake configuration because the CMake build system needs to be untangled, properly packaged, etc. But that is difficult if everything has to be mirrored to `autoconf` and maintained. At some point if the autotools system can be deprecated, there are a lot of simplifications we can make with CMake, including eliminating the need for custom shell scripts.

> To answer the question, the only libraries we build/install (as final products) are the libnetcdf library and the filter interface libraries, located in liblib/ and plugins/, respectively. We also build executables, located in ncdump/, ncgen/, ncgen3/. The other directories prefaced with lib are components of the final libnetcdf.a/libnetcdf.so, and are gated by options at configure time.

For now, this info is extremely helpful so on top-level `CMakeLists.txt` the targets should be:
```cmake
add_library(netcdf)
add_executable(ncdump)
add_executable(ncgen)
add_executable(ncgen3)
```
and then everything is just `target_sources` on these.

> filter interface libraries

Not sure what you mean by this. Are those proper libraries/executables? I looked at `liblib`, but that only defines `libnetcdf`, and `plugins` seems to involve some optional executables?

Reply:
I appreciate your explanation and your understanding; it is indeed a chicken-and-egg issue.  There has been conversation already around deprecating the autotools-based builds, but the impact on existing workflows in the community would be tremendous, in a way that is hard to develop an approach towards.  The joys of a broadly-used, 30+ years old project XD.  Logistics aside, I'm on board.

The `plugins` directory generates `interface libraries` used as part of `libhdf5`'s plugin architecture; for example, if you want to use a custom compressor with `libhdf5`-based storage,  you can't call the compression library directly, you have to have an "interface library".  These are what are built in `plugins/`; libraries that allow for a broader set of compression options.